File: GrayCode.java
public class Solution {
	public List<Integer> grayCode(int n) {
		int current = 0;
		List<Integer> result = new ArrayList<Integer>();
		result.add(current);
		int flip = 0;
		while(flip < n) {
			for(int i = result.size() - 1; i >= 0; i--) {
				result.add(result.get(i) | (1 << flip));
			}
			flip++;
		}
		return result;
	}
}

File: GrayCode.cc
class Solution {
    public:
        vector<int> grayCode(int n) {
            vector<int> result;
            int number = 0;
            int counter = 0;
            result.push_back(number);
            while(counter < n) {
                for (int i = result.size() - 1; i >= 0; i--) {
                    int new_elem = result[i] ^ (1 << counter);
                    result.push_back(new_elem);
                }
                counter++;
            }

            return result;
        }
};
